src/brod_transaction.erl

```diff
@@ -0,0 +1,343 @@
+-type key() :: brod:key().
+-type offset() :: kpro:offset().
+-type offsets_to_commit() :: kpro:offsets_to_commit().
+-type partition() :: kpro:partition().
+-type produce_ack_cb() :: brod:produce_ack_cb().
+-type topic() :: kpro:topic().
+-type transaction() :: pid().
+-type transactional_id() :: kpro:transactional_id().
+-type txn_ctx() :: kpro:txn_ctx().
+-type value() :: brod:value().
+
+
+-record(state,
+        { client_pid       :: client()
+        , context          :: txn_ctx()
+        , producers        :: map()
+        , producers_config :: config()
+        }).
+
+-type state() :: #state{}.
+
+init({Client, TxId, Config}) ->
+  ClientPid = pid(Client),
+  erlang:process_flag(trap_exit, true),
+  {ok, CTX} = make_txn_context(ClientPid, TxId),
+  {ok, #state{ client_pid = ClientPid
+             , context = CTX
+             , producers_config = Config
+             , producers = #{}
+             }}.
+
+%%% @private
+handle_call({add_offsets, ConsumerGroup, Offsets}, _From,
+            #state{ client_pid = Client
+                  , context = CTX
+                  } = State) ->
+  {ok, {{Host, Port}, _}} = brod_client:get_group_coordinator(Client, ConsumerGroup),
+  {ok, Conn} = brod_client:get_connection(Client, Host, Port),
+
+  %% before adding the offset we need to let kafka know we are going to commit
+  %% the offsets.
+  ok = kpro:txn_send_cg(CTX, ConsumerGroup),
+
+  %% actually commit the offset
+  Resp = kpro:txn_offset_commit(Conn, ConsumerGroup, CTX, Offsets),
+  {reply, Resp, State};
+
+
+%% @private
+handle_call({get_producer, Topic, Partition}, _From, #state{} = OldState) ->
+  State = may_add_producer(Topic, Partition, OldState),
+  {reply, get_producer(Topic, Partition, State), State};
+
+%% @private
+handle_call(commit, _From, #state{context = CTX} = State) ->
+  ok = kpro:txn_commit(CTX),
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(terminate, _From, State) ->
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(abort, _From,
+            #state{context = CTX} = State) ->
+  ok = kpro:txn_abort(CTX),
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call({abort, Timeout}, _From,
+            #state{context = CTX} = State) ->
+  ok = kpro:txn_abort(CTX, #{timeout => Timeout}),
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(stop, _From, #state{} = State) ->
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(Call, _From, #state{} = State) ->
+  {reply, {error, {unsupported_call, Call}}, State}.
+
+%% @private
+handle_cast(_Cast, #state{} = State) ->
+  {noreply, State}.
+
+%% @private
+handle_info(_Call, State) ->
+  {noreply, State}.
+
+%% @private
+code_change(_OldVsn, #state{} = State, _Extra) ->
+  {ok, State}.
+
+terminate(_Reason, _State) -> ok.
+
+
+%% @see start_link/2
+-spec new(pid(), config()) -> {ok, transaction()}.
+new(ClientPid, ProducerConfig) ->
```

Comment:
maybe return an `opaque()` transaction context like
`{ok, #{ client => ClientPid, transaction => TxnPid}}`, so the produce calls can lookup partition leader connections from client.